### ISSUE TITLE
Crash After First SleepEx With MASM

### ISSUE THREAD

**Issue by fin3ss3g0d** [Idov31/Cronos#3]
Curious why you are using NASM if you are already building this project inside of Visual Studio? I tried to convert your NASM code into MASM (it should be a direct translation):

```
REDACTED
```

I am calling it the same as your example from the C code but it is crashing after the first SleepEx. Where did you get the RSP calculations for the CONTEXT structures? How can I debug what is causing the issue after the first SleepEx?

**Comment by Idov31** [comment]
Hey,
For your first question - the project wasn't originally written as a VS project (see #1 ) so nasm was the preferred choice (matter of convenient).

Regarding your other questions, to debug it I recommend throwing it to IDA / Windbg / x64dbg or any other preferred debugger, put a breakpoint on the start of the rop and debug it line by line. We did the rsp calculation by hand and on a brief look it isn't clear what causing the problem.

**Comment by fin3ss3g0d** [comment]
Thanks for your quick reply! I have played with it more. So, I created this function that will simulate calling `SleepEx(1000, TRUE)` four times and excluded setting up the CONTEXT structures, just so I could test and see if the `SleepEx` ROP was working properly:

```
REDACTED
```

I placed a breakpoint on `SleepEx` and it indeed was called four times without any crashes or access violations. So, this would mean that the `SleepEx` ROP and the shadow space fixer are working properly. Would this mean that the RSP calculations for the CONTEXT structures could be off or have issues?

**Comment by fin3ss3g0d** [comment]
I found the problem. If I put a breakpoint on VirtualProtect it would cause an access violation. It probably messed up parts of the stack. Once I removed it everything worked properly. I'd still like to talk to you offline about how you calculated those RSP values for the CONTEXT structures in case you wanted to use different APIs for the ROP chain. I love how this example uses waitable timers, this work is AWESOME and thank you for sharing!
